### DISCUSSION TITLE
failure prediction based on server data metrics & log file

### DISCUSSION THREAD

**Discussion by SelimBF** [online-ml/river#880] [Q&A] [answered]
hello hope this question is not  asked before ,

I don't know if River could match my need, I want to build a failure prediction model based on data metrics & log file(time serie data) relatime feeded and collected from Linux server  to predict faillure before they occur using unsepervised model .
could any one tell me what should I do , is River helpful or not in such a satuation

**Comment by raphaelsty** [comment, marked as answer]
Hi @SelimBF, I think it is a use case where River could be practical; River is optimized to provide predictions continuously, and the models can be updated. 

An outlier detection model may do the job such as [OneClassSVM](https://riverml.xyz/latest/api/anomaly/OneClassSVM/).

**Reply by SelimBF** [reply]
thanks @raphaelsty  for replying , could you please or any one give me an exemple or demo I start with ,actually Iam new in ML ,I know basics only

**Reply by raphaelsty** [reply]
Hi @SelimBF,

Here is a fictitious example of an outlier detection algorithm applied to server metrics. There is an outlier (number of connections higher than usual, ram usage higher than average). We could imagine raising an alert. However, if we want to build a model that anticipates server failures, the input data should reveal the failure.

```python
>>> from river import anomaly
>>> from river import compose
>>> from river import datasets
>>> from river import metrics
>>> from river import preprocessing

>>> X_y = [
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
...    ({"cpu_usage": 0.99, "ram_usage": 0.99, "n_process": 200, "elasticsearch_cpu_usage": 0.99, "incoming_connexions_3_minutes": 200}, 1), # OUTLIER
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
... ]

>>> model = anomaly.QuantileThresholder(
...     anomaly.OneClassSVM(nu=0.2),
...     q=0.995
... )

>>> auc = metrics.ROCAUC()

>>> for x, y in X_y:
...     score = model.score_one(x)
...     model = model.learn_one(x)
...     auc = auc.update(y, score)

>>> auc
```

In this example I give scores 0 (everything is ok) and 1 (anomaly) simply to evaluate the model's capacity. In practice we won't need these labels which are simply used to evaluate the model.

**Reply by raphaelsty** [reply]
I imagine you would have to build a small API on your server with a route that returns the metrics. On another much smaller machine, you could continuously update the model by requesting your server every x seconds / minutes.

**Reply by SelimBF** [reply]
I really appreciate your help @raphaelsty  thanks a lot

**Reply by raphaelsty** [reply]
Let me know if it works the way you want it to! :)

**Reply by SelimBF** [reply]
@raphaelsty  
it seem the model isn't defined : 

 model = anomaly.QuantileThresholder(
IndentationError: unexpected indent

**Reply by raphaelsty** [reply]
Here it should work, my pseudo code had syntax errors

```python
>>> from river import anomaly
>>> from river import compose
>>> from river import datasets
>>> from river import metrics
>>> from river import preprocessing

>>> X_y = [
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
...    ({"cpu_usage": 0.99, "ram_usage": 0.99, "n_process": 200, "elasticsearch_cpu_usage": 0.99, "incoming_connexions_3_minutes": 200}, 1), # OUTLIER
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
...    ({"cpu_usage": 0.9, "ram_usage": 0.3, "n_process": 200, "elasticsearch_cpu_usage": 0.5, "incoming_connexions_3_minutes": 10}, 0),
... ]

>>> model = anomaly.QuantileThresholder(
...     anomaly.OneClassSVM(nu=0.2),
...     q=0.995
... )

>>> auc = metrics.ROCAUC()

>>> for x, y in X_y:
...     score = model.score_one(x)
...     model = model.learn_one(x)
...     auc = auc.update(y, score)

>>> auc
```

**Reply by SelimBF** [reply]
@raphaelsty  it return : 

ROCAUC: -0.00%

Process finished with exit code 0

This is ok I think !!

**Reply by raphaelsty** [reply]
Process finished with exit code 0 means that it run well with PyCharm. I hope it will work well on your data. Ideally, the AUC score would be 1. :)

**Reply by SelimBF** [reply]
so @raphaelsty if I got your point : I need to create a small rest API  that feed this model with metrics data collected from my server(CPU,memory.disk...etc) then call the API directly into my code to make it online ML

**Reply by raphaelsty** [reply]
Exactly! Ideally, your model could also be stored on an API 2, which regularly calls your server's API 1 to update the model. Then, you could call your model's API 2 to get predictions or raise an alert. :) I still recommend that you check that your model gives consistent results select the right metrics to raise the proper alert before deploying it in production.

**Reply by SelimBF** [reply]
@raphaelsty u are amazing bro , many thanks again

**Reply by SelimBF** [reply]
I will keep u updated bro

**Reply by SelimBF** [reply]
@raphaelsty  currently I am working routing log from server to SVM model using MQTT as I found that's the best match tool to be used in such situation  ...actually don't know if this concept logically can be realized because there no exemple that I can't start with,
if you suggest any exemple from your knowledge I will really appriciate that
